fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingBase.java

```diff
@@ -823,6 +823,12 @@ protected CompletableFuture<Void> iterateAllRanges(List<Object> additionalLogMes
                 getRunner().getExecutor());
     }
 
+    protected static boolean allRangesExhausted(Tuple cont, Tuple end) {
```

Comment:
In each case this function is used, it seems to be baking in an implicit assumption that `end` will always be `null`, as otherwise, the iteration of the loops won't ever stop, but I'm not sure that that's true in all cases

Reply:
I believe that the new condition effectively always assumes it is building the entire range (which is I suppose the "implicit assumption" that I thought it was "baking in"). That's wrong in the cases where we only want to build a range, though it sounds like we're trying to get rid of that in favor of an alternative approach anyway. But if that's the plan, I'd rather we just remove and deprecate those endpoints, possibly waiting until we have a better story around parallel index builds.

Now that I'm looking closer, I think the _original_ condition was also wrong, though maybe in a different way (I think). I'm mainly looking at `buildRangeOnly`, which is supposed to take a range, and ensure that that whole range is built. Before this PR, that code:

1. Reads the range set to compute unbuilt ranges in the given range then
2. Looks in the results and builds the range over multiple transactions until
3. It builds all the way until it reaches `rangeEnd`

However, `rangeEnd` there is a range from the range set, meaning it effectively only builds the first (contiguous) range of data from the target range. The desired behavior, I think, is that it build all the way until it gets to the end of the target range. So, for `buildRangeOnly`, I think a better condition is something like "until the tuple given matches a tuple constructed from `rangeEndBytes` (or is `null` if `rangeEndBytes` is null)".

Reply:
(assume:
'old path' = `IndexingByRecords` indexer - which was kept "as is" as much as possible. I'll be happy to change it as little as possible until reaching the stage of deleting unused code. 
'new path' = `IndexingMultiTargetByRecords`, 'IndexingByIndex`, `IndexingScrub*` indexers - a simplified path with no endpoint)

I believe that the current stop condition of the 'new path' was, indeed, a bug. My bug. It was harmless because these new indexers could not never have caused or 'inherit' a fragmented rangeSet. The name of the new stop condition - `allRangesExhausted` - should clarify that it return true if no more missing ranges exist, which is what these new indexers expect (I think). When the time comes, a different stop condition (`thisRangeIsExhausted`?) can be used to support a single range indexing. 
Although the 'old path' is hard to follow, I believe that its stop condition is fine because we do have some fragmented indexing tests. While developing, these tests had immediately failed when `IndexingMultiTargetByRecords` was defaulted to index single targets.
I'm not sure that I understood your last paragraph. IIUIC, `buildRangeOnly`'s return value is the stop condition of `iterateAllRanges`'s loop. For the current 'new path' indexers it returns true until `allRangesExhausted` - which means either continue indexing the same range or move on to the next unbuilt one. Am I missing something?